### PR TITLE
Internal Events API

### PR DESCRIPTION
**Pull request by APickledWalrus**
## Introduction

This PR aims to refactor and expand upon the existing [Script Event API](https://github.com/SkriptLang/Skript/blob/e2928ceae90c94beabe51240358d401b22524fba/src/main/java/org/skriptlang/skript/lang/script/Script.java#L161-L220).

This PR does NOT implement backwards compatibility of the existing Script Event API. The API was marked as experimental, and I do not believe it has widely been used. However, backwards compatibility may be implemented if we see it as important.

## [Event Registry](https://github.com/SkriptLang/Skript/blob/3d11e8b6ecf4ec39f83c9748da810bd4f35e5375/src/main/java/org/skriptlang/skript/util/EventRegistry.java)

A generic container for event instances. Events may be registered/unregistered. This simply serves for standardizing the event registration process across Skript. Below is an example of registering an event on ScriptLoader:
```java
ScriptLoader.eventRegistry().register(ScriptLoader.ScriptUnloadEvent, (parser, script) -> {
		myLogger.log("The script '" + script + "' has been unloaded");
});
```

## Events

Events have been added that can have handlers registered with ScriptLoader (global) and specific Script instances. To explain further, the example above registered an event with the ScriptLoader that logs when a Script has been unloaded. It would also be possible to register a handler for that event on a specific Script instance (to detect when that Script has been unloaded):

```java
someScript.eventRegistry().register(ScriptLoader.ScriptUnloadEvent, (parser, script) -> {
		// logic for when this script is unloaded (e.g. cleaning up some data)
});
```
### [ScriptPreInitEvent](https://github.com/SkriptLang/Skript/blob/3d11e8b6ecf4ec39f83c9748da810bd4f35e5375/src/main/java/ch/njol/skript/ScriptLoader.java#L1087-L1102)

This event is a replacement for the Bukkit-based [PreScriptLoadEvent](https://github.com/SkriptLang/Skript/blob/3d11e8b6ecf4ec39f83c9748da810bd4f35e5375/src/main/java/ch/njol/skript/events/bukkit/PreScriptLoadEvent.java). This event triggers right before the collection of Configs (pre-parsed script files) are loaded into actual Script instances. The collection is modifiable which allows the prevention of certain Configs from loading into Scripts.

### [ScriptInitEvent](https://github.com/SkriptLang/Skript/blob/3d11e8b6ecf4ec39f83c9748da810bd4f35e5375/src/main/java/ch/njol/skript/ScriptLoader.java#L1104-L1119)
This event is triggered right after a Script has been loaded into the ScriptLoader, but not yet fully initialized.

### [ScriptLoadEvent](https://github.com/SkriptLang/Skript/blob/3d11e8b6ecf4ec39f83c9748da810bd4f35e5375/src/main/java/ch/njol/skript/ScriptLoader.java#L1121-L1137)
This event is triggered right after a Script has finished loading. That is, all of its Structures have finished loading.

### [ScriptUnloadEvent](https://github.com/SkriptLang/Skript/blob/3d11e8b6ecf4ec39f83c9748da810bd4f35e5375/src/main/java/ch/njol/skript/ScriptLoader.java#L1139-L1154)
This event is triggered right before a Script is unloaded in the ScriptLoader. All data/Structures are still available.

### [ScriptActivityChangeEvent](https://github.com/SkriptLang/Skript/blob/3d11e8b6ecf4ec39f83c9748da810bd4f35e5375/src/main/java/ch/njol/skript/lang/parser/ParserInstance.java#L557-L576)
This event is triggered when the activity status of a ParserInstance changes (resulting in a Script being made active/inactive). This would trigger through the usage of `ParserInstance#setActive(Script)` and `ParserInstance#setInactive()`. This could be useful for managing data on Scripts that should reset when it is not active in a ParserInstance.

---
**Target Minecraft Versions:** Any
**Requirements:** None
**Related Issues:** N/A

### CODE REVIEW

**Comment by TheLimeGlass** [comment]
2.7?

File: src/main/java/org/skriptlang/skript/api/event/EventRegister.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	 * @param type The type of events to get.
+	 * @return An unmodifiable subset (of the specified type) of this register's events
+	 */
+	@Unmodifiable
+	@SuppressWarnings("unchecked")
+	public <T extends E> Set<T> getEvents(Class<T> type) {
+		return Collections.unmodifiableSet(
+			(Set<T>) events.stream()
+				.filter(event -> type.isAssignableFrom(event.getClass()))
+				.collect(Collectors.toSet())
```
Comment (TheLimeGlass) [inline src/main/java/org/skriptlang/skript/api/event/EventRegister.java:81]:
```suggestion
					.filter(event -> type.isAssignableFrom(event.getClass()))
					.collect(Collectors.toSet())
```

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

**Review by kiip1 — COMMENTED** [review]
What if we allow for `EventRegister`s to define a parent `EventRegister`? This would let `Script`'s event register to define `ScriptLoader`'s event register as parent (= get events will include the parent, so duplicate code can be prevented). This would require `ScriptLoaderEvent` to extend `ScriptEvent` (because of generics).

Example:
```java
// Old
eventRegister.getEvents(ScriptLoaderEvent.ScriptUnloadEvent.class)
		.forEach(eventHandler -> eventHandler.onUnload(parser, script));
script.getEventRegister().getEvents(ScriptLoaderEvent.ScriptUnloadEvent.class)
		.forEach(eventHandler -> eventHandler.onUnload(parser, script));

// New
script.getEventRegister().getEvents(ScriptLoaderEvent.ScriptUnloadEvent.class)
		.forEach(eventHandler -> eventHandler.onUnload(parser, script));
```

Let me know if you have any questions :)

**Comment by APickledWalrus** [comment]
> What if we allow for `EventRegister`s to define a parent `EventRegister`?

Discussed on Discord a little, but this might be difficult for ScriptEvents as ScriptLoader will eventually get unique events that only apply to it. This class *will* be extendable by developers though if they would have a case for this kind of functionality.

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/org/skriptlang/skript/util/EventRegister.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * EventRegisters are generic containers for registering events.
+ * They are used across Skript to provide additional API with consistent organization.
+ * @param <E> The class representing the type of events this register will hold.
+ */
+public class EventRegister<E> {
+
+	private final Set<E> events = new HashSet<>(5);
```
Comment (TheLimeGlass) [inline src/main/java/org/skriptlang/skript/util/EventRegister.java:35]: Why do we need to predefine the size here? Leave it default?

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

**Comment by sovdeeth** [comment]
conflicts need resolving!

**Review by sovdeeth — APPROVED** [review]

**Pull request merged**